File: frontend/src/lib/types/course.ts
export interface Course {
  id: string | null;
  lmsId: string | null;
  lmsProvider: string;
  name: string;
  code: string;
  instructor: string | null;
  description: string | null;
  semester: string | null;
  year: string | null;
}

export interface Module {
  id: string | null;
  lmsId: string | null;
  courseId: string;
  name: string;
  description: string | null;
  itemsCount: number;
  moduleItems: ModuleItem[];
  prerequisites: string[] | null;
}

export interface ModuleItem {
  id: string | null;
  lmsId: string | null;
  lmsContentId: string | null;
  name: string;
  moduleId: string;
  courseId: string;
  type: string;
  url: string | null;
  position: number | null;
}

export interface FlashcardDeck {
  id: string | null;
  courseId: string;
  moduleIds: string[];
  name: string;
  description: string | null;
  itemsCount: number;
  flashcardItems: FlashcardItem[];
}

export interface FlashcardItem {
  id: string | null;
  question: string;
  answer: string;
}

export interface Quiz {
  id: string | null;
  courseId: string;
  moduleIds: string[];
  name: string;
  description: string | null;
  itemsCount: number;
  previousAttempts: number;
  previousScore: number;
  quizItems: Question[];
}

export interface Question {
  id: string | null;
  question: string;
  options: string[];
  correctOption: number;
  explanation: string | null;
}

export interface FileDetails {
  id: string | null;
  canvasId: string | null;
  name: string;
  size: number;
  type: string;
  url: string;
}

File: backend/models/course.py
from pydantic import BaseModel, Field
from typing import Optional, List
from datetime import datetime

class CourseBase(BaseModel):
    name: str
    code: str
    description: Optional[str] = None
    instructor: Optional[str] = None
    semester: Optional[str] = None
    year: Optional[str] = None
    lms_id: Optional[str] = None
    lms_provider: Optional[str] = None

class CourseCreate(CourseBase):
    pass

class CourseUpdate(BaseModel):
    name: Optional[str] = None
    code: Optional[str] = None
    description: Optional[str] = None
    instructor: Optional[str] = None
    semester: Optional[str] = None
    year: Optional[str] = None

class Course(CourseBase):
    id: str
    created_at: datetime
    updated_at: datetime
    
    class Config:
        from_attributes = True

class Module(BaseModel):
    id: Optional[str] = None
    lms_id: Optional[str] = None
    course_id: str
    name: str
    description: Optional[str] = None
    items_count: Optional[int] = None
    prerequisites: Optional[str] = None

class ModuleItem(BaseModel):
    id: Optional[str] = None
    lms_id: Optional[str] = None
    name: str
    module_id: str
    course_id: str
    type: str
    url: Optional[str] = None
    lms_content_id: Optional[str] = None
    position: Optional[int] = None

class FileDetails(BaseModel):
    id: Optional[str] = None
    canvas_id: Optional[str] = None
    name: str
    url: str
    size: int
    type: str
    is_valid_for_processing: bool
    validation_error: Optional[str] = None
    file_type_name: Optional[str] = None
